File: scan.c
/*
  Copyright 2002-2004 John Plevyak, All Rights Reserved
*/

#include "d.h"

int scan_buffer(d_loc_t* aloc, D_State* parse_state, ShiftResult* results)
{
    d_loc_t loc = *aloc, last_loc = *aloc;
    char* s = loc.s;
    int col = loc.col, line = loc.line;
    int nresults = 0, i = 0, j;
    D_Shift **shift = NULL, **shift_diff = 0;

    switch (parse_state->scanner_size)
    {
        case 1:
        {
            /* all matches */
            SB_uint8* st = (SB_uint8*) parse_state->scanner_table;
            SB_trans_uint8* tst =
                (SB_trans_uint8*) parse_state->transition_table;
            uint8 state = 0, last = state, prev = state;
            uint8 c;
            uint32 sb, so;
            c = (uint8) *s++;
            while ((
                state =
                    st[state].scanner_block[(sb = (c >> SCANNER_BLOCK_SHIFT))]
                                           [(so = c & SCANNER_BLOCK_MASK)]))
            {
                state -= 1;
                if (prev && parse_state->accepts_diff)
                {
                    shift_diff =
                        parse_state
                            ->accepts_diff[tst[prev].scanner_block[sb][so]];
                    for (; *shift_diff; shift_diff++)
                    {
                        results[nresults].loc = loc;
                        results[nresults++].shift = *shift_diff;
                    }
                }
                prev = state;
                if (c == '\n')
                {
                    line++;
                    col = 0;
                }
                else
                    col++;
                loc.s = s;
                loc.line = line;
                loc.col = col;
                if (st[state].shift)
                {
                    last = state;
                    last_loc = loc;
                }
                c = (uint8) *s++;
            }
            shift = st[last].shift;
            break;
        }
        case 2:
        {
            /* all matches */
            SB_uint16* st = (SB_uint16*) parse_state->scanner_table;
            SB_trans_uint16* tst =
                (SB_trans_uint16*) parse_state->transition_table;
            uint16 state = 0, last = state, prev = state;
            uint8 c;
            uint32 sb, so;
            c = (uint8) *s++;
            while ((
                state =
                    st[state].scanner_block[(sb = (c >> SCANNER_BLOCK_SHIFT))]
                                           [(so = c & SCANNER_BLOCK_MASK)]))
            {
                state -= 1;
                if (prev && parse_state->accepts_diff)
                {
                    shift_diff =
                        parse_state
                            ->accepts_diff[tst[prev].scanner_block[sb][so]];
                    for (; *shift_diff; shift_diff++)
                    {
                        results[nresults].loc = loc;
                        results[nresults++].shift = *shift_diff;
                    }
                }
                prev = state;
                loc.s = s;
                loc.line = line;
                loc.col = col;
                if (st[state].shift)
                {
                    last = state;
                    last_loc = loc;
                }
                if (c == '\n')
                {
                    line++;
                    col = 0;
                }
                else
                    col++;
                c = (uint8) *s++;
            }
            shift = st[last].shift;
            break;
        }
        case 4:
        {
            /* all matches */
            SB_uint32* st = (SB_uint32*) parse_state->scanner_table;
            SB_trans_uint32* tst =
                (SB_trans_uint32*) parse_state->transition_table;
            uint32 state = 0, last = state, prev = state;
            uint8 c;
            uint32 sb, so;
            c = (uint8) *s++;
            while ((
                state =
                    st[state].scanner_block[(sb = (c >> SCANNER_BLOCK_SHIFT))]
                                           [(so = c & SCANNER_BLOCK_MASK)]))
            {
                state -= 1;
                if (prev && parse_state->accepts_diff)
                {
                    shift_diff =
                        parse_state
                            ->accepts_diff[tst[prev].scanner_block[sb][so]];
                    for (; *shift_diff; shift_diff++)
                    {
                        results[nresults].loc = loc;
                        results[nresults++].shift = *shift_diff;
                    }
                }
                prev = state;
                loc.s = s;
                loc.line = line;
                loc.col = col;
                if (st[state].shift)
                {
                    last = state;
                    last_loc = loc;
                }
                if (c == '\n')
                {
                    line++;
                    col = 0;
                }
                else
                    col++;
                c = (uint8) *s++;
            }
            shift = st[last].shift;
            break;
        }
    }
    if (shift)
    {
        for (; *shift; shift++)
        {
            results[nresults].loc = last_loc;
            results[nresults++].shift = *shift;
        }
    }
    if (nresults)
    {
        int longest = 0;
        char* end = results[nresults - 1].loc.s;
        if (parse_state->scan_kind == D_SCAN_LONGEST)
            longest = 1;
        if (parse_state->scan_kind == D_SCAN_MIXED)
        {
            for (i = nresults - 1; i >= 0; i--)
            {
                if (results[i].loc.s < end)
                    break;
                if (results[i].shift->shift_kind == D_SCAN_LONGEST)
                    longest = 1;
            }
        }
        if (longest)
        {
            /* keep only 'longest' */
            i = 0;
            for (j = 0; j < nresults; j++)
            {
                if (results[j].loc.s == end ||
                    results[j].shift->shift_kind == D_SCAN_TRAILING)
                {
                    if (i != j)
                        results[i] = results[j];
                    i++;
                }
            }
            nresults = i;
        }
        else if (parse_state->scan_kind == D_SCAN_MIXED)
        {
            /* only keep non-longest */
            for (j = i; j >= 0; j--)
                if (results[j].shift->shift_kind != D_SCAN_LONGEST)
                {
                    if (i != j)
                        results[i] = results[j];
                    i--;
                }
            nresults = nresults - i - 1;
            if (i != -1)
                memmove(&results[0],
                        &results[i + 1],
                        nresults * sizeof(results[0]));
        }
    }
    return nresults;
}

File: rust/dparser_lib/src/scan.rs
//! `scan.rs`
//! Native Rust scanner loop. Reads memory byte-by-byte navigating the multi-dimensional
//! deterministic finite automata embedded in the `D_State` table exactly replicating `scan.c`.

use crate::binary_format::{D_SCAN_LONGEST, D_SCAN_MIXED, D_Shift};
use crate::types::Loc;

pub const SCANNER_BLOCKS: usize = 4;
pub const SCANNER_BLOCK_SHIFT: u8 = 6;
pub const SCANNER_BLOCK_MASK: u8 = 63;

/// Trait to safely traverse generic scanner sizes (uint8, uint16, uint32)
pub trait ScannerTable {
    unsafe fn next_state(&self, state: usize, sb: usize, so: usize) -> usize;
    unsafe fn shifts(&self, state: usize) -> *mut *mut D_Shift;
}

#[repr(C)]
struct SBUint8 {
    shift: *mut *mut D_Shift,
    scanner_block: [*mut u8; SCANNER_BLOCKS],
}

impl ScannerTable for *const SBUint8 {
    unsafe fn next_state(&self, state: usize, sb: usize, so: usize) -> usize {
        let st = self.add(state);
        let block_ptr = (*st).scanner_block[sb];
        if block_ptr.is_null() {
            0
        } else {
            *block_ptr.add(so) as usize
        }
    }
    unsafe fn shifts(&self, state: usize) -> *mut *mut D_Shift {
        (*self.add(state)).shift
    }
}

#[repr(C)]
struct SBUint16 {
    shift: *mut *mut D_Shift,
    scanner_block: [*mut u16; SCANNER_BLOCKS],
}

impl ScannerTable for *const SBUint16 {
    unsafe fn next_state(&self, state: usize, sb: usize, so: usize) -> usize {
        let st = self.add(state);
        let block_ptr = (*st).scanner_block[sb];
        if block_ptr.is_null() {
            0
        } else {
            *block_ptr.add(so) as usize
        }
    }
    unsafe fn shifts(&self, state: usize) -> *mut *mut D_Shift {
        (*self.add(state)).shift
    }
}

#[repr(C)]
struct SBUint32 {
    shift: *mut *mut D_Shift,
    scanner_block: [*mut u32; SCANNER_BLOCKS],
}

impl ScannerTable for *const SBUint32 {
    unsafe fn next_state(&self, state: usize, sb: usize, so: usize) -> usize {
        let st = self.add(state);
        let block_ptr = (*st).scanner_block[sb];
        if block_ptr.is_null() {
            0
        } else {
            *block_ptr.add(so) as usize
        }
    }
    unsafe fn shifts(&self, state: usize) -> *mut *mut D_Shift {
        (*self.add(state)).shift
    }
}

#[repr(C)]
struct SBTransUint8 {
    scanner_block: [*mut u8; SCANNER_BLOCKS],
}
#[repr(C)]
struct SBTransUint16 {
    scanner_block: [*mut u16; SCANNER_BLOCKS],
}
#[repr(C)]
struct SBTransUint32 {
    scanner_block: [*mut u32; SCANNER_BLOCKS],
}

#[derive(Clone, Debug)]
pub struct ShiftResult {
    pub loc: Loc,
    pub shift: crate::grammar::GrammarShift,
}

pub fn scan_buffer(
    input: &[u8],
    starting_loc: Loc,
    parse_state: &crate::grammar::GrammarState,
) -> Vec<ShiftResult> {
    let mut results = Vec::new();
    let mut loc = starting_loc;
    let mut last_loc = starting_loc;

    let mut shift_ptr: *mut *mut D_Shift = std::ptr::null_mut();

    if parse_state.scanner_table == 0 {
        // Fallback for missing scanner table mappings explicitly native
    } else {
        let mut execute_scan =
            |st: &dyn ScannerTable, get_trans: &dyn Fn(usize, usize, usize) -> usize| {
                let mut state = 0;
                let mut last_state = 0;
                let mut prev = 0;

                // Loop purely through characters matching the boundary of the `input` slice
                // `loc.s` maps to actual parsed offset locally safely
                while loc.s < input.len() {
                    let c = input[loc.s];
                    let sb = (c >> SCANNER_BLOCK_SHIFT) as usize;
                    let so = (c & SCANNER_BLOCK_MASK) as usize;

                    unsafe {
                        let next = st.next_state(state, sb, so);
                        if next == 0 {
                            break;
                        }
                        state = next - 1;

                        // Process differential `accepts_diff` table branches if embedded natively inside state node
                        if prev > 0 && parse_state.accepts_diff != 0 {
                            let trans_idx = get_trans(prev, sb, so);
                            let accepts_diff_raw = parse_state.accepts_diff as *mut *mut *mut D_Shift;
                            let mut diff_ptr = *accepts_diff_raw.add(trans_idx);
                            while !diff_ptr.is_null() && !(*diff_ptr).is_null() {
                                
                                let c_diff = &**diff_ptr;
                                let spec_raw: isize = std::mem::transmute(c_diff.speculative_code);

                                let safe_shift = crate::grammar::GrammarShift {
                                    symbol: c_diff.symbol,
                                    shift_kind: c_diff.shift_kind,
                                    op_assoc: c_diff.op_assoc,
                                    op_priority: c_diff.op_priority,
                                    term_priority: c_diff.term_priority,
                                    action_index: c_diff.action_index,
                                    speculative_code: spec_raw as i32,
                                };

                                results.push(ShiftResult {
                                    loc,
                                    shift: safe_shift,
                                });
                                diff_ptr = diff_ptr.add(1);
                            }
                        }

                        prev = state;

                        if c == b'\n' {
                            loc.line += 1;
                            loc.col = 0;
                        } else {
                            loc.col += 1;
                        }
                        loc.s += 1;

                        let current_shift = st.shifts(state);
                        if !current_shift.is_null() {
                            last_state = state;
                            last_loc = loc;
                        }
                    }
                }

                unsafe {
                    shift_ptr = st.shifts(last_state);
                }
            };

        unsafe {
            match parse_state.scanner_size {
                1 => {
                    let st = parse_state.scanner_table as *const SBUint8;
                    let tst = parse_state.transition_table as *const SBTransUint8;
                    let trans_fn = |p: usize, sb: usize, so: usize| {
                        if (*tst.add(p)).scanner_block[sb].is_null() {
                            0
                        } else {
                            *(*tst.add(p)).scanner_block[sb].add(so) as usize
                        }
                    };
                    execute_scan(&st, &trans_fn);
                }
                2 => {
                    let st = parse_state.scanner_table as *const SBUint16;
                    let tst = parse_state.transition_table as *const SBTransUint16;
                    let trans_fn = |p: usize, sb: usize, so: usize| {
                        if (*tst.add(p)).scanner_block[sb].is_null() {
                            0
                        } else {
                            *(*tst.add(p)).scanner_block[sb].add(so) as usize
                        }
                    };
                    execute_scan(&st, &trans_fn);
                }
                4 => {
                    let st = parse_state.scanner_table as *const SBUint32;
                    let tst = parse_state.transition_table as *const SBTransUint32;
                    let trans_fn = |p: usize, sb: usize, so: usize| {
                        if (*tst.add(p)).scanner_block[sb].is_null() {
                            0
                        } else {
                            *(*tst.add(p)).scanner_block[sb].add(so) as usize
                        }
                    };
                    execute_scan(&st, &trans_fn);
                }
                _ => {
                    return Vec::new();
                }
            }
        }
    }

    if !shift_ptr.is_null() {
        unsafe {
            let mut ptr = shift_ptr;
            while !(*ptr).is_null() {
                let c_shift = &**ptr;
                let spec_raw: isize = std::mem::transmute(c_shift.speculative_code);

                let safe_shift = crate::grammar::GrammarShift {
                    symbol: c_shift.symbol,
                    shift_kind: c_shift.shift_kind,
                    op_assoc: c_shift.op_assoc,
                    op_priority: c_shift.op_priority,
                    term_priority: c_shift.term_priority,
                    action_index: c_shift.action_index,
                    speculative_code: spec_raw as i32,
                };

                results.push(ShiftResult {
                    loc: last_loc,
                    shift: safe_shift,
                });
                ptr = ptr.add(1);
            }
        }
    }

    // Filter Longest/Mixed natively
    if !results.is_empty() {
        let mut longest = false;
        let scan_kind_longest = D_SCAN_LONGEST as u8;
        let scan_kind_mixed = D_SCAN_MIXED as u8;

        let end_idx = results.last().unwrap().loc.s;
        if parse_state.scan_kind == scan_kind_longest {
            longest = true;
        } else if parse_state.scan_kind == scan_kind_mixed {
            for res in results.iter().rev() {
                if res.loc.s < end_idx {
                    break;
                }
                if res.shift.shift_kind == scan_kind_longest {
                    longest = true;
                }
            }
        }

        if longest {
            results.retain(|r| r.loc.s == end_idx);
        }
    }

    results
}
